model_compression_toolkit/data_generation/keras/optimization_utils.py

```diff
@@ -0,0 +1,436 @@
+
+        total_mean /= self.n_batches
+        total_second_moment /= self.n_batches
+        return tf.expand_dims(input=total_mean, axis=0), tf.expand_dims(input=total_second_moment, axis=0)
+
+    def compute_bn_loss(self,
+                        input_imgs: tf.Tensor,
+                        batch_stats_holder: BatchStatsHolder,
+                        activation_extractor: KerasActivationExtractor) -> tf.Tensor:
+        """
+        Compute the batch norm alignment loss.
+
+        Args:
+            input_imgs (Any): the input images.
+            batch_stats_holder (BatchStatsHolder): batch stats holder.
+            activation_extractor (ActivationExtractor): extractor for layer activations.
+
+        Returns:
+            Any: the computed batch norm alignment loss.
+        """
+
+        self.all_imgs_stats_holder.update_batch_stats(batch_stats_holder=batch_stats_holder,
+                                                      input_imgs=input_imgs,
+                                                      activation_extractor=activation_extractor)
+
+        # Initialize variables for accumulating mean and variance differences
+        total_bn_loss, total_mean_diff, total_var_diff = 0, 0, 0
+
+        # Iterate over each BN layer
+        layer_weights = self.bn_layer_weights_fn(orig_bn_stats_holder=self.orig_bn_stats_holder)
+
+        for layer_name in self.orig_bn_stats_holder.get_bn_layer_names():
+            # Get the layer weight for the current BN layer
+            layer_weight = layer_weights.get(layer_name)
+
+            # Get the mean and variance from the original BN statistics
+            bn_layer_mean = self.orig_bn_stats_holder.get_mean(bn_layer_name=layer_name)
+            bn_layer_var = self.orig_bn_stats_holder.get_var(bn_layer_name=layer_name)
+
+            # Get the mean and variance from the current batch's statistics
+            if self.use_all_data_stats:
+                # If using all data statistics, retrieve the accumulated statistics from all batches
+                imgs_layer_mean, imgs_layer_var = self.get_accumulated_stats_per_layer(layer_name=layer_name)
+            else:
+                # Otherwise, retrieve the statistics from the current batch
+                imgs_layer_mean, imgs_layer_var = (
+                    self.all_imgs_stats_holder.get_stats(batch_stats_holder=batch_stats_holder, layer_name=layer_name))
+
+            bn_layer_std = tf.sqrt(bn_layer_var + self.eps)
+            imgs_layer_std = tf.sqrt(imgs_layer_var + self.eps)
+
+            total_bn_loss += layer_weight * self.bn_loss_fn(bn_mean=bn_layer_mean,
+                                                            input_mean=imgs_layer_mean,
+                                                            bn_std=bn_layer_std,
+                                                            input_std=imgs_layer_std)
+        return total_bn_loss
+
+    def compute_output_loss(self,
```

Comment:
output_loss_multiplier should be outside this function, in the main "data_generation" function